File: Chapter 06/JS/menu.js
const readline = require('readline');
const path = require('path');
const fs = require('fs');

// Load architectural data separated from logic
const config = require('./examples.json');
const examples = config.examples;

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function clearScreen() {
    console.clear();
}

function askQuestion(query) {
    return new Promise(resolve => rl.question(query, resolve));
}

async function main() {
    while (true) {
        clearScreen();
        console.log(`=== ${config.title} ===\n`);

        for (const [key, example] of Object.entries(examples)) {
            console.log(`${key}. ${example.name}`);
        }
        console.log("\nType 'exit' to quit.");
        
        const choice = (await askQuestion("\nEnter your choice: ")).trim().toLowerCase();
        
        if (choice === 'exit') {
            console.log("Exiting menu...");
            rl.close();
            break;
        } 
        
        if (examples[choice]) {
            const relativePath = examples[choice].path;
            const name = examples[choice].name;
            const isServer = examples[choice].isServer;
            
            const scriptPath = path.join(__dirname, relativePath);
            
            clearScreen();
            
            if (!fs.existsSync(scriptPath)) {
                console.log(`[ERROR] Could not find file: ${scriptPath}`);
                await askQuestion("\nPress Enter to return to the menu...");
                continue; 
            }

            console.log(`--- Running ${name} ---\n`);
            
            try {
                // Clear cache to allow multiple runs in one session
                delete require.cache[require.resolve(scriptPath)];
                const Demo = require(scriptPath);
                
                if (Demo && typeof Demo.run === 'function') {
                    // Call the static run method
                    const activeInstance = await Demo.run();
                    
                    if (isServer) {
                        await askQuestion("\nPress Enter to stop the server and return to the main menu...");
                        if (activeInstance && typeof activeInstance.close === 'function') {
                            activeInstance.close();
                            console.log("Server shut down successfully.");
                        }
                    } else {
                        await askQuestion("\nPress Enter to return to the main menu...");
                    }
                } else {
                    console.log("[ERROR] Demo class or exported run() method not found.");
                    await askQuestion("\nPress Enter to return to the main menu...");
                }
            } catch (err) {
                console.log(`[ERROR] Execution failed: ${err.message}`);
                await askQuestion("\nPress Enter to return to the main menu...");
            }
        } else {
            console.log("Invalid choice. Please try again.");
            await askQuestion("\nPress Enter to continue...");
        }
    }
}

main();

File: Chapter 06/Python/menu.py
import json
import os
import importlib
import asyncio
import inspect

def clear_screen():
    os.system('cls' if os.name == 'nt' else 'clear')

def main():
    config_path = 'examples.json'

    # BULLETPROOF CHECK: Does the config file exist?
    if not os.path.exists(config_path):
        print(f"[ERROR] {config_path} not found!")
        return

    with open(config_path, 'r') as f:
        examples = json.load(f)

    while True:
        # clear_screen() # Keep clear_screen for better user experience
        print("=== Grokking Software Architecture Chapter 06: Python Examples ===\n")

        # Sort keys numerically for a clean menu
        keys = sorted(examples.keys(), key=lambda x: int(x))
        for key in keys: 
            print(f"{key}. {examples[key]['name']}")

        print("\nType 'exit' to quit.")
        choice = input("\nEnter your choice: ").strip()

        if choice.lower() == 'exit':
            break

        if choice in examples:
            selected = examples[choice]
            # clear_screen() # Keep clear_screen for better user experience
            print(f"--- Running {selected['name']} ---\n")

            try:
                module = importlib.import_module(selected['path'])
                demo_class = getattr(module, 'Demo', None)

                if demo_class and hasattr(demo_class, 'run'):
                    # Check if the example itself is an async function
                    if inspect.iscoroutinefunction(demo_class.run):
                        asyncio.run(demo_class.run())
                    else:
                        # Standard execution (Allows Uvicorn to manage its own event loop)
                        demo_class.run()
                else:
                    print(f"[ERROR] Could not find class 'Demo' or method 'run' in {selected['path']}")

            except Exception as e:
                print(f"[ERROR] Execution failed: {str(e)}")
            
            input("\nPress Enter to return to the main menu...")
        else:
            input("Invalid choice. Press Enter to try again...")

if __name__ == "__main__":
    # Don't use asyncio.run() so Uvicorn can take control
    main()